pkg/schedulingtypes/jobscheduler.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schedulingtypes
+
+import (
+	"fmt"
+
+	fedschedulingv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/scheduling/v1alpha1"
+	fedclientset "github.com/kubernetes-sigs/federation-v2/pkg/client/clientset/versioned"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	. "github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util/planner"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	kubeclientset "k8s.io/client-go/kubernetes"
+	crclientset "k8s.io/cluster-registry/pkg/client/clientset/versioned"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-sigs/federation-v2/pkg/schedulingtypes/adapters"
+)
+
+const (
+	JobSchedulingKind = "JobSchedulingPreference"
+)
+
+func init() {
+	RegisterSchedulingType(JobSchedulingKind, NewJobScheduler)
+}
+
+type JobScheduler struct {
+	plugin    *Plugin
+	fedClient fedclientset.Interface
+}
+
+func NewJobScheduler(fedClient fedclientset.Interface, kubeClient kubeclientset.Interface, crClient crclientset.Interface, fedNamespace, clusterNamespace, targetNamespace string, federationEventHandler, clusterEventHandler func(pkgruntime.Object), handlers *ClusterLifecycleHandlerFuncs) Scheduler {
+	scheduler := &JobScheduler{}
+	scheduler.fedClient = fedClient
+	adapter := adapters.NewFederatedJobAdapter(fedClient)
+	tc, err := fedClient.CoreV1alpha1().FederatedTypeConfigs(util.DefaultFederationSystemNamespace).Get(JobTypeName, metav1.GetOptions{})
+	if err != nil {
+		glog.Errorf("Failed to get APIResource for %s: %v", JobTypeName, err)
```

Comment:
If this error occurs, the scheduler will error out on line 59.  I think this highlights the fragility of the current mechanism for enabling scheduling types (all or nothing).  If a type config for a type targeted by a given scheduler is either missing or not enabled, consider failing scheduler start.  

A future refinement could be adding scheduler management to the federation type config controller (or adding a similar controller dedicated to scheduling) so that a given scheduler could be started/stopped in response to the addition/deletion/modification of federated type configs that scheduling depends on.